### PR TITLE
Allow custom HTTP headers in REST client, log correlation ID

### LINKED ISSUE #673: Allow specifying query parameters and HTTP headers in REST client calls
It should be possible to extend the PowerAuth REST client calls by specifying custom query parameters and HTTP headers for the use case when additional data should be sent with the request data (e.g. for auditing purposes).

### LINKED ISSUE #684: Add support for logging using correlation ID with MDC
Once https://github.com/wultra/powerauth-server/issues/673 is implemented, we should add a configurable option to include correlation ID from HTTP headers in server logs.

Following changes are required:
- add configuration property to enable logging with correlation ID, e.g. `*.correlation-header.enabled=true`
- add configuration property for correlation ID header name, e.g. `*.correlation-header.name=Correlation-ID`
- add the correlation ID value into MDC context
- document how to set up logging with correlation headers using Spring configuration properties

### PR DESCRIPTION
**Pull request by romanstrobl**
- Fix #673: Allow specifying query parameters and HTTP headers in REST client calls
- Fix #684: Add support for logging using correlation ID with MDC

### CODE REVIEW

**Review by petrdvorak — APPROVED** [review]
Looks OK, please review the year in copyright headers and final keyword.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+@ConditionalOnProperty(
+        value = "powerauth.service.correlation-header.enabled",
+        havingValue = "true"
+)
+public class InterceptorConfiguration implements WebMvcConfigurer {
+
+    @Value("${powerauth.service.correlation-header.name:Correlation-ID}")
+    private String correlationHeaderName;
+
+    @Value("${powerauth.service.correlation-header.value.validation-regexp:[a-zA-Z0-9]}")
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java]: Yes, this is already fixed, I will just change the max. length from 128 to 1024.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Component
+@ConditionalOnProperty(
+        value = "powerauth.service.correlation-header.enabled",
+        havingValue = "true"
+)
+public class InterceptorConfiguration implements WebMvcConfigurer {
+
+    @Value("${powerauth.service.correlation-header.name:Correlation-ID}")
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java]: Already fixed.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    private String correlationHeaderName;
+
+    @Value("${powerauth.service.correlation-header.value.validation-regexp:[a-zA-Z0-9]}")
+    private String correlationHeaderValueValidation;
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        HandlerInterceptor httpHeaderInterceptor = new HttpHeaderInterceptor(correlationHeaderName, correlationHeaderValueValidation);
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/InterceptorConfiguration.java:48]: Added.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/interceptor/HttpHeaderInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex) throws Exception {
+        MDC.remove(correlationHeaderName);
+    }
+
+    private String getCorrelationId(HttpServletRequest request) {
+        String headerValue = request.getHeader(correlationHeaderName);
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/interceptor/HttpHeaderInterceptor.java:57]: Added.

File: powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java [hunk 5]
```diff
@@ -90,25 +95,61 @@ public PowerAuthRestClient(String baseUrl, PowerAuthRestClientConfiguration conf
...
+    public void setQueryParams(MultiValueMap<String, String> queryParams) {
+        this.queryParams = queryParams;
+    }
+
+    /**
+     * Set HTTP headers.
+     * @param httpHeaders HTTP headers.
+     */
+    public void setHttpHeaders(MultiValueMap<String, String> httpHeaders) {
+        this.httpHeaders = httpHeaders;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java]: This is already fixed.

**Pull request merged**